maintenance/uploadXmlDumpToSpectrumS3Bucket.php

```diff
@@ -0,0 +1,128 @@
+		$this->addOption( 'bucketName', 'S3 bucket name', true, true, 'b' );
+		$this->addOption( 'saveChanges', 'Get XML dump and save it in s3 bucket', false, false, 'd' );
+	}
+
+	public function execute() {
+
+		global $wgMaxShellTime, $wgMaxShellFileSize, $wgDumpsDisabledWikis;
+
+		$AWSAccessKey = $this->getOption( 'AWSAccessKey' );
+		$AWSSecretKey = $this->getOption( 'AWSSecretKey' );
+		$bucketName =  $this->getOption( 'bucketName' );
+		$saveChanges = $this->hasOption( 'saveChanges' );
+
+		$range = array();
+		$range[] = "city_public = 1";
+
+		if( !empty( $wgDumpsDisabledWikis ) && is_array( $wgDumpsDisabledWikis ) ) {
+			$range[] = 'city_id NOT IN (' . implode( ',', $wgDumpsDisabledWikis ) . ')';
+		}
+
+		$wgMaxShellTime = 0;
+		$wgMaxShellFileSize = 0;
+
+		$dbw = Wikifactory::db( DB_MASTER );
+		$sth = $dbw->select(
+			array( "city_list" ),
+			array( "city_id", "city_dbname" ),
+			$range,
+			__METHOD__,
+			array( "ORDER BY" => "city_id" )
+		);
+
+		$s3 = new S3( $AWSAccessKey, $AWSSecretKey );
+
+		while( $row = $dbw->fetchObject( $sth ) ) {
+			$this->output( "Preparing XML dump for {$row->city_id} - {$row->city_dbname}\n" );
+			if ( $saveChanges ) {
+				$this->runBackup( $bucketName, $s3, $row,
+					sprintf( "%s/%s_pages_current.xml.7z", sys_get_temp_dir(), $row->city_dbname ),
+					[ '--current' ] );
+				$this->output( "Done, XML dump saved in S3 bucket\n" );
+
+			}
+		}
+
+	}
+
+	private function putToAmazonS3( string $sPath, bool $bPublic, string $sMimeType, string $bucketName, $s3 )  {
+
+		$size = filesize( $sPath );
+
+		S3::setExceptions( true );
+
+		$file = fopen( $sPath, 'rb' );
+		$resource = S3::inputResource( $file, $size );
+		$remotePath = DumpsOnDemand::getPath( basename( $sPath ) );
+
+		$s3->putObject(
+			$resource,
+			$bucketName,
+			$remotePath,
+			$bPublic ? S3::ACL_PUBLIC_READ : S3::ACL_PRIVATE,
+			[],
+			[
+				'Content-Disposition' => 'attachment',
+				'Content-Type' => $sMimeType
+			]
+		);
+	}
+
+	private function runBackup( $bucketName, $s3, $row, $path, array $args = [] ) {
+		global $IP, $wgWikiaLocalSettingsPath, $options;
+
+		$server = wfGetDB( DB_SLAVE, 'dumps', $row->city_dbname )->getProperty( "mServer" );
+		$cmd = implode( ' ', array_merge( [
+			"SERVER_ID={$row->city_id}",
+			"php -d display_errors=1",
+			"{$IP}/maintenance/dumpBackup.php",
+			"--conf {$wgWikiaLocalSettingsPath}",
+			"--xml",
+			"--quiet",
+			"--server=$server",
+			"--output=".DumpsOnDemand::DEFAULT_COMPRESSION_FORMAT.":{$path}"
+		], $args ) );
+
+		$cmd .= ' 2>&1';
+		wfShellExec( $cmd, $status );
+
+		if ( $status === 0 ) {
+			try {
+				$this->putToAmazonS3( $path, !isset( $options[ "hide" ] ),  MimeMagic::singleton()->guessMimeType(
+					$path
+				), $bucketName, $s3 );
+				$this->output("XML dump was successful\n");
+			} catch( S3Exception $ex ) {
+				$this->output( "Failed to put XML dump for {$row->city_id} ({$row->city_dbname}) to bucket {$ex}\n" );
+				exit( 1 );
```

Comment:
Are we sure we want to stop execution on first failure? It might be better to keep it going at look at all the failures after it's finished.

Reply:
that's what we agreed on with @harnash - to stop the script after one S3 failure since next wikis would probably failed too. After that we can run script again starting from specific wiki, which reminded me: 78e1cfe